preload.js

```diff
@@ -42,26 +42,32 @@ function renderTray() {
 	logo.src = 'favicon.ico';
 }
 
-function addUnreadEvent() {
+function appStateChange(event, state) {
+	if(['OPENING', 'DISCONNECTED', 'TIMEOUT'].includes(state)) {
```

Comment:
Here's the code responsible to check the app state (phone disconnected, no internet). `OPENING` is when internet is disconnected, `TIMEOUT` is when the phone is disconnected, and I'm not sure what triggers `DISCONNECTED`, I'm including it purely because its name.

Reply:
Here's the thing though, after prolonged testing I found out that `OPENING` seem to randomly happen even though whatsapp isn't actually disconnected. But this doesn't stay long, the state then changes to `PAIRING` and `CONNECTED` quickly. I guess putting the check on a timeout is going to work, but I wonder if there's another property at `window.Store.AppState` that's more accurate, `stream` is a promising property. More digging is needed.

Reply:
Okay. Then for now we will just rely on Opening, Disconnected and Timeout and later if we find that any case is still being left out we will fix that. Meanwhile, I will also test different cases and see if anything is being left out.

Reply:
So using a timeout then? Because currently it's randomly showing notification

Reply:
huh... that random notification might be annoying for some users. So, I guess will just remove the TIMEOUT element from the array so you don't have to create another pull request. Seems fine?

Reply:
No, read again. `OPENING` seem to randomly happen even though whatsapp isn't actually disconnected. But this doesn't stay long, the state then changes to `PAIRING` and `CONNECTED` quickly. I suggest changing the code to something like this:
```js
function appStateChange(event, state) {
    if(['OPENING', 'DISCONNECTED', 'TIMEOUT'].includes(state)) {
        setTimeout(() => {
            if (state === window.Store.AppState.state) {
                new Notification('WALC disconnected', {
                    body: "Please check your connection.",
                    icon: "favicon.ico"
                });
            }
        }, 3000);
    }
}```

I haven't tested this though

Reply:
That's great way I also thought of something like this but only for Timeout State but I guess we can do it for all the states. I will test this and fix any errors. Although if before timeout the state again changes like from Timeout to disconnected then it might be a problem so I will also test that the new state is not in the above array. Something like this:
```javascript
function appStateChange(event, state) {
    if(['OPENING', 'DISCONNECTED', 'TIMEOUT'].includes(state)) {
        setTimeout(() => {
            if (['OPENING', 'DISCONNECTED', 'TIMEOUT'].contains(window.Store.AppState.state)) {
                new Notification('WALC disconnected', {
                    body: "Please check your connection.",
                    icon: "favicon.ico"
                });
            }
        }, 3000);
    }
}
```

Reply:
`TIMEOUT` state only happens when phone is disconnected, after testing I have found no issue with that state, the problematic one is `OPENING`. If the state changes again during timeout, it will simply call `appStateChange` again thus starting another timeout. If you check the state to the array, then the notification will appear twice because the check will return true on both timeout.

Reply:
Oh okay. I got your point. I will just use the `setTimeout` block provided by you then. so we don't show multiple Notifications.